### PR TITLE
Adds callback select system

### PR DESCRIPTION
**Pull request by MrsTonedOne**
This system lets you run a group of callbacks asynchronously, returning once all of them finish, without having to wait for each callback to finish before running the next one.

I made it in my bicon pr, but i don't need it anymore, so i removed it from there and put it here.

### CODE REVIEW

**Comment by RemieRichards** [comment]
still hate the name

File: code/datums/callback.dm [hunk 1]
```diff
@@ -100,3 +100,60 @@
...
+
+/datum/callback_select/proc/invoke_callback(index, datum/callback/callback, list/callback_args, savereturn = TRUE)
+	set waitfor = FALSE
+	if (!callback || !istype(callback))
+		//This check only exists because the alternative is callback_select would block forever if given invalid data
+		CRASH("invalid callback passed to invoke_callback")
+	if (!length(callback_args))
+		callback_args = list()
+	pendingcount++
+	debug_usr("calling callback")
```
Comment (JJRcop) [inline code/datums/callback.dm]: Debug call

File: code/datums/callback.dm [hunk 2]
```diff
@@ -100,3 +100,60 @@
...
+	var/rtn = callback.Invoke(arglist(callback_args))
+	debug_usr("callback returned")
```
Comment (JJRcop) [inline code/datums/callback.dm]: Debug call

File: code/datums/callback.dm [hunk 3]
```diff
@@ -100,3 +100,60 @@
...
+		return
+	if (!callback_args)
+		callback_args = list()
+
+	callback_args.len = count
+
+	var/datum/callback_select/CS = new(count, savereturns)
+	for (var/i in 1 to count)
+		CS.invoke_callback(i, callbacks[i], callback_args[i], savereturns)
+	debug_usr("starting callbacks: [CS.pendingcount]")
```
Comment (JJRcop) [inline code/datums/callback.dm]: Debug call

File: code/datums/callback.dm [hunk 4]
```diff
@@ -100,3 +100,60 @@
...
+	while(CS.pendingcount)
+		debug_usr("callbacks: [CS.pendingcount]")
```
Comment (JJRcop) [inline code/datums/callback.dm]: Debug call

**Review by JJRcop — CHANGES_REQUESTED** [review]
Code looks good to me, there are some lingering debug calls to a proc that doesn't exist.

**Review by JJRcop — APPROVED** [review]

**Review by Cyberboss — APPROVED** [review]

**Pull request merged**